### ISSUE TITLE
[meta] Optimisation strategies

### ISSUE THREAD

**Issue by andreiglingeanu** [ThemeFuse/Unyson#1920]
It's not a secret that Unyson has it's own speed flaws. They really drive crazy most of us and they should be fixed. I'll try to add here most known ones and I'd like to see posts for the outside, maybe something that other developers found out.

<hr>

_small notes_:
1. Keep it focused. One post should one real, argumented issue. Further discussions on those should be done in a separate issue for each of them, if they're really big. Trash will be filtered.
2. What part of the system it really affects? Admin side? Frontend side? The speed load of the pages? Which extension is responsible for that? Be descriptive.
3. This is not a thread about suggesting features. It is about finding problems out and solving them.
4. Pull requests are accepted for each problem in this thread. Be sure to link us to them.

**Comment by andreiglingeanu** [comment]
I'll start on with one.

<hr>

**WordPress Shortcodes** extension is great. But it introduces a BIG [`wp_localize_script`](https://github.com/ThemeFuse/Unyson-WP-Shortcodes-Extension/blob/v1.0.0/static.php#L14-L22).  It contains `options`, `values`, `title` for **each** shortcode (even the ones defined in your themes) and it includes it on every admin page. That should be fixed.

**Comment by thednp** [comment]
As of now, I'm happy with how it all works in terms of performance, because I know how it was before.

I would personally focus on small little things we all need, I'm not going to list my list here, an easy [search](https://github.com/ThemeFuse/Unyson/issues/created_by/thednp) would reveal them.

**Comment by andreiglingeanu** [comment]
@thednp I'm almost done about the `WP Shortcodes` fix. Some other really important problems to come here.

By the way, I've also though about the non-jQuery themes. Too bad lots of plugins/widgets actually depend on jQuery.

**Comment by andreiglingeanu** [comment]
Next one on the list, monster [**Icon V2**](http://manual.unyson.io/en/latest/options/built-in-option-types.html#icon-v2) option type.

<hr>

The problem:

1. Once again, we have really big [wp_localize_script()](https://files.slack.com/files-pri/T1NK80NLU-F2414PQ1H/menubar_and_view-source_localhost_3003_wp-admin_admin_php_page_fw-settings_and_localhost_3003__page_id_35_tab-1.png?pub_secret=8aa364e338) data.
2. Really slow first-time modal render. REALY SLOOOOW.

**Comment by danyj** [comment]
This one also, 
https://github.com/ThemeFuse/Unyson/issues/1491

the more tabs you add , the slower first save is.

**Comment by thednp** [comment]
> By the way, I've also though about the non-jQuery themes. Too bad lots of plugins/widgets actually depend on jQuery.

@andreiglingeanu the future starts here, my theme currently working on is based on this concept, it only needs jQuery for the Gmaps and Contact form shortcodes (also Contact Form 7), but I will get to those later. Everything else is covered by [bootstrap.native](https://github.com/thednp/bootstrap.native/) and some other small scripts.

**Comment by andreiglingeanu** [comment]
@thednp Good work on getting bootstrap to work without `jQuery`. The thing is that **Bootstrap 4** is already coming, but they still [keep depending](https://github.com/twbs/bootstrap/blob/v4.0.0-alpha.3/js/src/alert.js) on jQuery, which is sad. But I think they'll happily accept some lightweight alternative like `Zepto` or even `jqlite` from `angular.js`.

> I was actually thinking of a way to mimic the dynamic loading of scripts for jQuery so that it's loaded only if required by the shortcodes. I know this sounds SCI FI, but I don't believe it to be impossible.

Totally not impossible. We're actually doing something like this, but a bit more smarter about our frontend components (shortcodes, if you will).

**Comment by andreiglingeanu** [comment]
Okay, `icon-v2` is down.

**Comment by andreiglingeanu** [comment]
A whole suite of pull requests were just merged that will prevent page builder from doing `wp_localize_script` for each shortcode or item type he needs.

https://github.com/ThemeFuse/Unyson-Forms-Extension/pull/39
https://github.com/ThemeFuse/Unyson-PageBuilder-Extension/pull/70
https://github.com/ThemeFuse/Unyson-Shortcodes-Extension/pull/69

This also means that you have any shortcodes data you may need, without the need of Page Builder. This means that we are one step closer to making Unyson shortcodes available anywhere (maybe even frontend one day).

**Comment by andreiglingeanu** [comment]
Also, Unyson Shortcodes are now loaded in a fancy new [modal](https://i.gyazo.com/42b4ad509dc2ee6be711575c8c70bcfb.gif).

**Comment by danyj** [comment]
https://github.com/ThemeFuse/Unyson/issues/2037

**Comment by andreiglingeanu** [comment]
Update on https://github.com/ThemeFuse/Unyson/issues/1920#issuecomment-246717904: We still have `wp_localize_script` loaded in Page Builder - we can't easily delay the initialisation of PB because of a poor [integration](https://github.com/ThemeFuse/Unyson-PageBuilder-Extension/blob/v1.6.3/includes/page-builder/static/js/editor_integration.js) with WordPress Post editor. Still marked as TODO.

**Comment by andreiglingeanu** [comment]
https://github.com/ThemeFuse/Unyson/issues/1581#issuecomment-256254274

**Comment by ivandotv** [comment]
Check this out.
For front-end I'm not using files that create back-end options.
So in case of `customizer.php` Inside that file immediately at the top I'm checking to see if I'm in admin or or if wp customizer iz active, if that is false I immediately return, and options in that file are not parsed.
In frontend when trying to get the options from the customzier via `fw_get_db_customizer_option($option_id, $default_value = null)`  as default value **never** pass in `null` so unyson never tries to get the value from options in the file.
Of course that means I get the default options from somewhere else.
So I have a global array that holds default values, so unyson options, and front end options can get the values from that array.
I hope this is clear.
Basically unyson options are never parsed when front-end of the site is loaded, but unyson does check for the values that I need, but it only checks the DB not the "hard-coded" options.
Also when creating "hard-coded" options we can also use that *global* variable.

**Comment by andreiglingeanu** [comment]
@ivandotv We probably need to create some sort of a smart cache for options, that's a better way than having a _global_ variable for storing this state. If I remember correctly, @mowshon was thinking about ways to cache options files. And yes, I am aware about that this will give a big performance boost.

cc. @GheorgheP

**Comment by ivandotv** [comment]
Yeah, `global` variable was the easiest and fastest way to get this going. It can easily be a class that holds the data.

**Comment by ivandotv** [comment]
Also, I have custom `framework dir` and I could also respond to the `fw_framework_customizations_dir_rel_path` with and empty string. And there are no errors, but I don't see much perfromance gain from that?
